plugins/inventory/lxd.py

```diff
@@ -0,0 +1,1090 @@
+                    for interface_family in self.data['inventory'][container_name].get('network_interfaces')[interface]:
+                        try:
+                            if self.ip_in_subnetwork(interface_family['address'], self.groups[group_name].get('attribute')):
+                                self.inventory.add_child(group_name, container_name)
+                        except ValueError:
+                            # catch ipv4/ipv6 matching incompatibility errors
+                            continue
+
+    def build_inventory_groups_os(self, group_name):
+        """create group by attribute: os
+
+        Args:
+            str(group_name): Group name
+        Kwargs:
+            Noneself.data['inventory'][container_name][interface]
+        Raises:
+            None
+        Returns:
+            None"""
+        # maybe we just want to expand one group
+        if group_name not in self.inventory.groups:
+            self.inventory.add_group(group_name)
+
+        gen_containers = [
+            container_name for container_name in self.inventory.hosts.keys()
+            if 'ansible_lxd_os' in self.inventory.get_host(container_name).get_vars().keys()]
+        for container_name in gen_containers:
+            if self.groups[group_name].get('attribute').lower() == self.inventory.get_host(container_name).get_vars().get('ansible_lxd_os'):
+                self.inventory.add_child(group_name, container_name)
+
+    def build_inventory_groups_release(self, group_name):
+        """create group by attribute: release
+
+        Args:
+            str(group_name): Group name
+        Kwargs:
+            None
+        Raises:
+            None
+        Returns:
+            None"""
+        # maybe we just want to expand one group
+        if group_name not in self.inventory.groups:
+            self.inventory.add_group(group_name)
+
+        gen_containers = [
+            container_name for container_name in self.inventory.hosts.keys()
+            if 'ansible_lxd_release' in self.inventory.get_host(container_name).get_vars().keys()]
+        for container_name in gen_containers:
+            if self.groups[group_name].get('attribute').lower() == self.inventory.get_host(container_name).get_vars().get('ansible_lxd_release'):
+                self.inventory.add_child(group_name, container_name)
+
+    def build_inventory_groups_profile(self, group_name):
+        """create group by attribute: profile
+
+        Args:
+            str(group_name): Group name
+        Kwargs:
+            None
+        Raises:
+            None
+        Returns:
+            None"""
+        # maybe we just want to expand one group
+        if group_name not in self.inventory.groups:
+            self.inventory.add_group(group_name)
+
+        gen_containers = [
+            container_name for container_name in self.inventory.hosts.keys()
+            if 'ansible_lxd_profile' in self.inventory.get_host(container_name).get_vars().keys()]
```

Comment:
`var in somedict.keys()` does not work in Python 3, as `.keys()` no longer returns a `set`.
```suggestion
            if 'ansible_lxd_profile' in list(self.inventory.get_host(container_name).get_vars().keys())]
```

Reply:
same problem in multiple places

Reply:
Hi russoz,

you are right the return-type is dict_keys and not list.
I use this 18 times. Mostly listgenerators or for loops an it works with python3.
Maybe i dont understand the problem.
Can you enplane why a cast to list is necessary?
e.g.:

```
Python 3.8.6 (default, Jan 27 2021, 15:42:20) 
[GCC 10.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> a = {'a':'b', 'c':'d'}
>>> a
{'a': 'b', 'c': 'd'}
>>> a.keys()
dict_keys(['a', 'c'])
>>> list(a.keys())
['a', 'c']
>>> for item in a.keys():
...     print(item)
... 
a
c
```

Reply:
My mistake - I thought the operator `in` would fail with the `dict_keys` object but that's not true. what actually **cannot** happen is trying to access an element by index:
```python
Python 3.8.5 (default, Jan 27 2021, 15:41:15) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> d = dict(a=1, b=2, c=3)
>>> d
{'a': 1, 'b': 2, 'c': 3}
>>> d.keys()
dict_keys(['a', 'b', 'c'])
>>> 'a' in d.keys()
True
>>> 'f' in d.keys()
False
>>> d.keys()[3]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'dict_keys' object is not subscriptable
```

Reply:
Instead of `a in d.keys()` or `a in list(d.keys())`, simply use `a in d`. That's more Pythonic **and** more efficient.